### PR TITLE
Support glTF Specular Glossiness Extension

### PR DESCRIPTION
**Pull request by OmarShehata**
This PR adds support for the [KHR_materials_pbrSpecularGlossiness glTF extension](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_pbrSpecularGlossiness/README.md). Resolves https://github.com/AnalyticalGraphicsInc/cesium/issues/6921.

I still have to write tests, but I'm opening it now to get some early feedback. 

It's currently implemented in `processPbrMetallicRoughness.js` because it seems implementation-wise the only difference between Metallic Roughness and Specular Glossiness is how the BRDF inputs are computed. It'll check for this extension and modify the relevant part of the shader, keeping the rest of it intact. 

Since `processPbrMetallicRoughness.js` would now be doing both metallic-roughness and specular-glossiness, should it be renamed to something more general like `processPbrMaterials.js` ? The other option would be to abstract away most of `processPbrMetallicRoughness.js` into a common class that could be shared. @emackey what do you think?

I added a test model with spec-gloss in `Specs/Data/Models/PBR/BoomBoxSpecularGlossiness/`. Here's a snippet to test it locally in Sandcastle, and to switch between the metallic-roughness and the specular-glossiness versions of the model.

```
var viewer = new Cesium.Viewer('cesiumContainer', {
    selectionIndicator : false,
});

viewer.scene.globe.enableLighting = true;
viewer.scene.globe.depthTestAgainstTerrain = true;

var position = new Cesium.Cartesian3(-1371108.1511167218, -5508684.080096612, 2901825);
var heading = Cesium.Math.toRadians(180);
var pitch = Cesium.Math.toRadians(2);
var roll = Cesium.Math.toRadians(-6);
var hpr = new Cesium.HeadingPitchRoll(heading, pitch, roll);
var orientation = Cesium.Transforms.headingPitchRollQuaternion(position, hpr);

var bareGLTF = viewer.entities.add({
    position : position,
    orientation : orientation,
    model : {
        uri : '../../../Specs/Data/Models/PBR/BoomBox/BoomBox.gltf',
        heightReference : Cesium.HeightReference.RELATIVE_TO_GROUND ,
        minimumPixelSize : 512,
        maximumScale : 100,
        scale : 8.0,
        runAnimations : false
    }
});

viewer.trackedEntity = bareGLTF;

var position2 = new Cesium.Cartesian3(-1371100.6511167218, -5508684.080096612, 2901825);

var specGlossGLTF = viewer.entities.add({
    position : position2,
    orientation : orientation,
    model : {
        uri :'../../../Specs/Data/Models/PBR/BoomBoxSpecularGlossiness/BoomBox.gltf',
        heightReference : Cesium.HeightReference.RELATIVE_TO_GROUND ,
        minimumPixelSize : 512,
        maximumScale : 100,
        scale : 8.0,
        runAnimations : false
    }
});


var options = [{
    text : 'glTF',
    onselect : function() {
        viewer.trackedEntity = bareGLTF;
    }
}, {
    text : 'glTF Spec Gloss',
    onselect : function() {
        viewer.trackedEntity = specGlossGLTF;
    }
}];

Sandcastle.addToolbarMenu(options);
```

For this model, they look identical which I think is intended. 

normal

Adding a `specularFactor: [1.0, 0.5, 0.5]` gives the specular a reddish tone:

reddish

For reference, [this is how Threejs implements it](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/GLTFLoader.js#L551).

The only other thing is that the spec says:

> In addition to the material properties, if a primitive specifies a vertex color using the attribute semantic property COLOR_0, then this value acts as an additional linear multiplier to diffuse.

@lilleyse do you know if this is something that's already handled?

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by lilleyse** [comment]
Very cool @OmarShehata!

> In addition to the material properties, if a primitive specifies a vertex color using the attribute semantic property COLOR_0, then this value acts as an additional linear multiplier to diffuse.
>
> @lilleyse do you know if this is something that's already handled?

Yeah we handle that already, for metallic roughness at least - https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/ThirdParty/GltfPipeline/processPbrMetallicRoughness.js#L529

File: Source/Scene/processPbrMetallicRoughness.js [hunk 1]
```diff
@@ -578,9 +634,17 @@ define([
...
 
             fragmentShader += '    vec3 f0 = vec3(0.04);\n';
-            fragmentShader += '    float alpha = roughness * roughness;\n';
-            fragmentShader += '    vec3 diffuseColor = baseColor * (1.0 - metalness) * (1.0 - f0);\n';
-            fragmentShader += '    vec3 specularColor = mix(f0, baseColor, metalness);\n';
+            if (useSpecGloss) {
+                fragmentShader += '    float alpha = pow((1.0 - glossiness), 2.0);\n';
+                fragmentShader += '    vec3 diffuseColor = diffuse.rgb * (1.0 - max(max(specular.r, specular.g), specular.b));\n';
+                fragmentShader += '    vec3 specularColor = specular;\n';
+                fragmentShader += '    float roughness = 1.0 - glossiness;\n';
```
Comment (emackey) [inline Source/Scene/processPbrMetallicRoughness.js]: Do this above the `alpha` calculation on line 638 above, then you can avoid the `pow` and just have `float alpha = roughness * roughness`.

File: Source/Scene/processPbrMetallicRoughness.js [hunk 2]
```diff
@@ -531,29 +549,67 @@ define([
...
+                        fragmentShader += '    glossiness *= u_glossinessFactor;\n';
+                    }
+                } else {
+                    if (defined(generatedMaterialValues.u_specularFactor)) {
+                        fragmentShader += '    vec3 specular = clamp(u_specularFactor, vec3(0.0), vec3(1.0));\n';
+                    } else {
+                        fragmentShader += '    vec3 specular = vec3(1.0);\n';
+                    }
+                    if (defined(generatedMaterialValues.u_glossinessFactor)) {
+                        fragmentShader += '    float glossiness *= clamp(u_glossinessFactor, 0.0, 1.0);\n';
```
Comment (emackey) [inline Source/Scene/processPbrMetallicRoughness.js]: You have `float glossiness *=` here, that doesn't compile.

File: Source/Scene/processPbrMetallicRoughness.js [hunk 3]
```diff
@@ -531,29 +549,67 @@ define([
...
+                    fragmentShader += '    vec3 specular = specularGlossiness.rgb;\n';
+                    fragmentShader += '    float glossiness = specularGlossiness.a;\n';
+                    if (defined(generatedMaterialValues.u_specularFactor)) {
+                        fragmentShader += '    specular *= u_specularFactor;\n';
+                    }
+                    if (defined(generatedMaterialValues.u_glossinessFactor)) {
+                        fragmentShader += '    glossiness *= u_glossinessFactor;\n';
+                    }
+                } else {
+                    if (defined(generatedMaterialValues.u_specularFactor)) {
```
Comment (emackey) [inline Source/Scene/processPbrMetallicRoughness.js]: eslint is complaining here, make sure you're set up to run the code style tools.

File: Source/Scene/processPbrMetallicRoughness.js [hunk 4]
```diff
@@ -531,29 +549,67 @@ define([
...
-                if (defined(generatedMaterialValues.u_metallicFactor)) {
-                    fragmentShader += '    float metalness = clamp(u_metallicFactor, 0.0, 1.0);\n';
+                if (defined(generatedMaterialValues.u_diffuseTexture)) {
+                    fragmentShader += '    vec4 diffuse = SRGBtoLINEAR4(texture2D(u_diffuseTexture, ' + v_texcoord + '));\n';
+                    if (defined(generatedMaterialValues.u_diffuseFactor)) {
+                        fragmentShader += '    diffuse *= u_diffuseFactor;\n';
+                    }
                 } else {
-                    fragmentShader += '    float metalness = 1.0;\n';
+                    if (defined(generatedMaterialValues.u_diffuseFactor)) {
```
Comment (emackey) [inline Source/Scene/processPbrMetallicRoughness.js]: eslint is complaining here, make sure you're set up to run the code style tools.

File: Source/Scene/processPbrMetallicRoughness.js (hunk 2 above)
Comment (emackey) [reply Source/Scene/processPbrMetallicRoughness.js]: I tried fixing this locally by removing the `*`.  This works and shows my SpecGlossVsMetalRough new test model correctly!

**Comment by emackey** [comment]
Need to merge in `gltf-update-pipeline` and resolve conflicts.

Left some comments above, otherwise looks great!

specglosstest

**Comment by OmarShehata** [comment]
Thanks for the review @emackey, and for the water bottle test case! 

It should be all good now. The test cases I added check for being able to render a model when all parameters are provided, when none are provided and when no textures are provided, so it would have caught the shader compile error earlier. 

I also renamed the class to `processPbrMaterials` because I think that better reflects what it does now.

**Comment by OmarShehata** [comment]
@emackey did you want to give this another look for the new commits? (Biggest change is adding tests and renaming the class).

**Comment by emackey** [comment]
Sure.  Can't today, will try to carve out time tomorrow.

**Comment by OmarShehata** [comment]
Thanks Ed! And yeah, I believe those are because the base branch `gltf-update-pipeline` doesn't have updated tests yet.

File: Source/Scene/processPbrMaterials.js [hunk 5]
```diff
@@ -107,16 +112,30 @@ define([
         var shaders = techniquesWebgl.shaders;
         var programs = techniquesWebgl.programs;
 
+        var useSpecGloss = isSpecularGlossinessMaterial(material);
+
         var uniformName;
         var pbrMetallicRoughness = material.pbrMetallicRoughness;
-        if (defined(pbrMetallicRoughness)) {
-            for (var parameterName in pbrMetallicRoughness) {
+        var parameterName;
```
Comment (lilleyse) [inline Source/Scene/processPbrMaterials.js]: Minor style note:

```
var uniformName;
var parameterName;
var pbrMetallicRoughness = material.pbrMetallicRoughness;
```

looks cleaner than

```
var uniformName;
var pbrMetallicRoughness = material.pbrMetallicRoughness;
var parameterName;
```

**Comment by lilleyse** [comment]
> And what is needed for AnalyticalGraphicsInc:gltf-update-pipeline to be merged into master?

There was one failing model (CesiumAir.gltf), but we identified the fix in gltf-pipeline and it just needs to be updated back in Cesium. I believe that was the last thing holding up the PR. @ggetz can confirm.

File: Source/Scene/processPbrMaterials.js [hunk 6]
```diff
@@ -88,7 +88,7 @@ define([
         return gltf;
     }
 
-    function isPbrMaterial(material) {
+    function isMetallicRoughnessMaterial(material) {
```
Comment (lilleyse) [inline Source/Scene/processPbrMaterials.js]: This should actually not be renamed. The function is meant to catch if PBR shaders need to be generated. It should be catching whether the material uses metallic roughness, spec gloss, or unlit. The purpose is to turn away materials that are using `KHR_techniques_webgl` or `KHR_materials_common` (1.0).

One way to test is with a material like:
```
 {
  "extensions": {
    "KHR_materials_pbrSpecularGlossiness": {}
  }
}
```

Comment (lilleyse) [reply Source/Scene/processPbrMaterials.js]: This check probably also belongs before the `hasPbrMetallicRoughness` check:

```
        // No need to create new techniques if they already exist,
        // the shader should handle these values
        if (hasExtension(gltf, 'KHR_techniques_webgl')) {
            return gltf;
        }
```

I feel like a lot of the material existence checking was a little messy before this PR and could use some cleanup.

Comment (OmarShehata) [reply Source/Scene/processPbrMaterials.js]: Ah, good catch! I think it's helpful to have those two functions separately, so I might just add:

```
function isPbrMaterial(material){
  return isMetallicRoughnessMaterial(material) || isSpecularGlossinessMaterial(material);
}

Comment (lilleyse) [reply Source/Scene/processPbrMaterials.js]: Sounds good - though make sure sure the extreme edge case works too. This should produce a default PBR shader.

```
"materials": [
  {
  }
]
```

Comment (OmarShehata) [reply Source/Scene/processPbrMaterials.js]: Wow, you're right. I didn't realize glTF materials were all PBR by default. That extreme case currently doesn't work in master. It should be good here now. 

So `isMetallicRoughnessMaterial()` isn't needed at all then. If any material is found, then it is metallic roughness, unless spec-gloss is also found.

Comment (lilleyse) [reply Source/Scene/processPbrMaterials.js]: Yeah it might not be needed at all.

**Comment by OmarShehata** [comment]
Thanks for the cleanup suggestions @lilleyse ! It should be good to merge now.

**Comment by OmarShehata** [comment]
@lilleyse have you gotten a chance to look at this yet/see if it's ready to merge?

File: Source/Scene/processPbrMaterials.js [hunk 7]
```diff
@@ -23,26 +23,26 @@ define([
...
+        var hasPbrMaterial = false;
+        ForEach.material(gltf, function(material) {
+            // All materials in glTF are PBR by default,
+            // so if any material is found, we should apply PBR.
+            hasPbrMaterial = true;
+        });
+
+        if (!hasPbrMaterial) {
+            return gltf;
+        }
```
Comment (lilleyse) [inline Source/Scene/processPbrMaterials.js]: Simpler to just return early if there are no materials.

**Comment by OmarShehata** [comment]
@lilleyse I simplified that check, but to make it compatible with both glTF 2.0 and 1.0 I had to make sure to check if it's an array vs an object.

**Comment by lilleyse** [comment]
@OmarShehata the 2.0 vs 1.0 distinction isn't necessary. The model will always be 2.0 coming into this function.

**Comment by OmarShehata** [comment]
Ah, that makes it even simpler. Should be good now.

**Comment by lilleyse** [comment]
Thanks @OmarShehata!

**Pull request merged**